File: apps/frontend/src/services/grpc/mapcode.service.ts
/**
 * MapCode Service Client (gRPC-Web)
 * ======================
 * Real gRPC client implementation for MapCodeService
 * Replaces stub implementation with actual backend calls
 * 
 * @author NyNus Development Team
 * @version 2.0.0 - Real gRPC Implementation
 * @created 2025-01-19
 */

/* eslint-disable @typescript-eslint/no-explicit-any */

// gRPC-Web imports
import { MapcodeServiceClient } from '@/generated/v1/MapcodeServiceClientPb';
import {
  MapCodeVersion as PbMapCodeVersion,
  MapCodeTranslation as PbMapCodeTranslation,
  HierarchyNavigation as PbHierarchyNavigation,
  HierarchyLevel as PbHierarchyLevel,
  StorageInfo as PbStorageInfo,
  CreateVersionRequest,
  GetVersionsRequest,
  GetActiveVersionRequest,
  SetActiveVersionRequest,
  DeleteVersionRequest,
  TranslateCodeRequest,
  TranslateCodesRequest,
  GetHierarchyNavigationRequest,
  GetStorageInfoRequest,
} from '@/generated/v1/mapcode_pb';
import { RpcError } from 'grpc-web';

// gRPC client utilities
import { getGrpcUrl } from '@/lib/config/endpoints';
import { getAuthMetadata } from './client';

// Frontend types (from mapcode-client.ts)
export interface MapCodeVersionData {
  id: string;
  version: string;
  name: string;
  description: string;
  filePath: string;
  isActive: boolean;
  createdBy: string;
  createdAt: Date;
  updatedAt: Date;
}

export interface MapCodeTranslationData {
  questionCode: string;
  translation: string;
  grade?: string;
  subject?: string;
  chapter?: string;
  level?: string;
  lesson?: string;
  form?: string;
}

export interface HierarchyLevelData {
  code: string;
  name: string;
  description: string;
}

export interface HierarchyNavigationData {
  questionCode: string;
  grade?: HierarchyLevelData;
  subject?: HierarchyLevelData;
  chapter?: HierarchyLevelData;
  level?: HierarchyLevelData;
  lesson?: HierarchyLevelData;
  form?: HierarchyLevelData;
  breadcrumbs: string[];
}

export interface StorageInfoData {
  totalVersions: number;
  maxVersions: number;
  availableSlots: number;
  warningLevel: number;
  isNearLimit: boolean;
  isAtLimit: boolean;
  warningMessage: string;
}

// ===== gRPC CLIENT INITIALIZATION =====

const GRPC_ENDPOINT = getGrpcUrl();
const mapcodeServiceClient = new MapcodeServiceClient(GRPC_ENDPOINT);

// ===== OBJECT MAPPERS =====

/**
 * Map protobuf MapCodeVersion to frontend MapCodeVersionData
 */
function mapVersionFromPb(pbVersion: PbMapCodeVersion): MapCodeVersionData {
  const versionObj = pbVersion.toObject();
  
  return {
    id: versionObj.id,
    version: versionObj.version,
    name: versionObj.name,
    description: versionObj.description,
    filePath: versionObj.filePath,
    isActive: versionObj.isActive,
    createdBy: versionObj.createdBy,
    createdAt: versionObj.createdAt?.seconds ? new Date(versionObj.createdAt.seconds * 1000) : new Date(),
    updatedAt: versionObj.updatedAt?.seconds ? new Date(versionObj.updatedAt.seconds * 1000) : new Date(),
  };
}

/**
 * Map protobuf MapCodeTranslation to frontend MapCodeTranslationData
 */
function mapTranslationFromPb(pbTranslation: PbMapCodeTranslation): MapCodeTranslationData {
  const translationObj = pbTranslation.toObject();
  
  return {
    questionCode: translationObj.questionCode,
    translation: translationObj.translation,
    grade: translationObj.grade || undefined,
    subject: translationObj.subject || undefined,
    chapter: translationObj.chapter || undefined,
    level: translationObj.level || undefined,
    lesson: translationObj.lesson || undefined,
    form: translationObj.form || undefined,
  };
}

/**
 * Map protobuf HierarchyLevel to frontend HierarchyLevelData
 */
function mapHierarchyLevelFromPb(pbLevel: PbHierarchyLevel | undefined): HierarchyLevelData | undefined {
  if (!pbLevel) return undefined;
  
  const levelObj = pbLevel.toObject();
  return {
    code: levelObj.code,
    name: levelObj.name,
    description: levelObj.description,
  };
}

/**
 * Map protobuf HierarchyNavigation to frontend HierarchyNavigationData
 */
function mapHierarchyNavigationFromPb(pbNavigation: PbHierarchyNavigation): HierarchyNavigationData {
  const navigationObj = pbNavigation.toObject();
  
  return {
    questionCode: navigationObj.questionCode,
    grade: navigationObj.grade ? mapHierarchyLevelFromPb(pbNavigation.getGrade()) : undefined,
    subject: navigationObj.subject ? mapHierarchyLevelFromPb(pbNavigation.getSubject()) : undefined,
    chapter: navigationObj.chapter ? mapHierarchyLevelFromPb(pbNavigation.getChapter()) : undefined,
    level: navigationObj.level ? mapHierarchyLevelFromPb(pbNavigation.getLevel()) : undefined,
    lesson: navigationObj.lesson ? mapHierarchyLevelFromPb(pbNavigation.getLesson()) : undefined,
    form: navigationObj.form ? mapHierarchyLevelFromPb(pbNavigation.getForm()) : undefined,
    breadcrumbs: navigationObj.breadcrumbsList || [],
  };
}

/**
 * Map protobuf StorageInfo to frontend StorageInfoData
 */
function mapStorageInfoFromPb(pbStorage: PbStorageInfo): StorageInfoData {
  const storageObj = pbStorage.toObject();
  
  return {
    totalVersions: storageObj.totalVersions,
    maxVersions: storageObj.maxVersions,
    availableSlots: storageObj.availableSlots,
    warningLevel: storageObj.warningLevel,
    isNearLimit: storageObj.isNearLimit,
    isAtLimit: storageObj.isAtLimit,
    warningMessage: storageObj.warningMessage,
  };
}

// ===== ERROR HANDLING =====

/**
 * Handle gRPC errors and convert to user-friendly messages
 */
function handleGrpcError(error: RpcError): string {
  console.error('gRPC Error:', error);
  switch (error.code) {
    case 3: return error.message || 'Dữ liệu không hợp lệ';
    case 5: return 'Không tìm thấy MapCode version';
    case 7: return 'Bạn không có quyền thực hiện thao tác này';
    case 14: return 'Dịch vụ tạm thời không khả dụng';
    case 16: return 'Vui lòng đăng nhập để tiếp tục';
    default: return error.message || 'Đã xảy ra lỗi không xác định';
  }
}

// ===== MAPCODE SERVICE IMPLEMENTATION =====

export const MapCodeService = {
  /**
   * Create a new MapCode version
   * Tạo version mới cho MapCode
   */
  createVersion: async (
    version: string,
    name: string,
    description: string,
    createdBy: string
  ): Promise<MapCodeVersionData> => {
    try {
      const request = new CreateVersionRequest();
      request.setVersion(version);
      request.setName(name);
      request.setDescription(description);
      request.setCreatedBy(createdBy);

      const response = await mapcodeServiceClient.createVersion(request, getAuthMetadata());
      const versionData = response.getVersion();
      
      if (!versionData) {
        throw new Error('No version returned from server');
      }

      return mapVersionFromPb(versionData);
    } catch (error) {
      const errorMessage = handleGrpcError(error as RpcError);
      throw new Error(errorMessage);
    }
  },

  /**
   * Get all versions with pagination
   * Lấy danh sách tất cả versions
   */
  getVersions: async (page: number = 1, pageSize: number = 10): Promise<{
    versions: MapCodeVersionData[];
    pagination: {
      currentPage: number;
      pageSize: number;
      totalPages: number;
      totalItems: number;
      hasNext: boolean;
      hasPrevious: boolean;
    };
  }> => {
    try {
      const request = new GetVersionsRequest();
      request.setPage(page);
      request.setPageSize(pageSize);

      const response = await mapcodeServiceClient.getVersions(request, getAuthMetadata());
      const versions = response.getVersionsList().map(mapVersionFromPb);
      const paginationResponse = response.getPagination();
      
      const totalPages = paginationResponse?.getTotalPages() || 1;
      const totalItems = paginationResponse?.getTotalCount() || 0;
      
      return {
        versions,
        pagination: {
          currentPage: page,
          pageSize,
          totalPages,
          totalItems,
          hasNext: page < totalPages,
          hasPrevious: page > 1,
        },
      };
    } catch (error) {
      const errorMessage = handleGrpcError(error as RpcError);
      throw new Error(errorMessage);
    }
  },

  /**
   * Get the currently active version
   * Lấy version đang active
   */
  getActiveVersion: async (): Promise<MapCodeVersionData> => {
    try {
      const request = new GetActiveVersionRequest();

      const response = await mapcodeServiceClient.getActiveVersion(request, getAuthMetadata());
      const version = response.getVersion();
      
      if (!version) {
        throw new Error('No active version found');
      }

      return mapVersionFromPb(version);
    } catch (error) {
      const errorMessage = handleGrpcError(error as RpcError);
      throw new Error(errorMessage);
    }
  },

  /**
   * Set a version as active
   * Đặt version làm active
   */
  setActiveVersion: async (versionId: string): Promise<string> => {
    try {
      const request = new SetActiveVersionRequest();
      request.setVersionId(versionId);

      const response = await mapcodeServiceClient.setActiveVersion(request, getAuthMetadata());
      const status = response.getStatus();

      return status?.getMessage() || 'Version activated successfully';
    } catch (error) {
      const errorMessage = handleGrpcError(error as RpcError);
      throw new Error(errorMessage);
    }
  },

  /**
   * Delete a version
   * Xóa version
   */
  deleteVersion: async (versionId: string): Promise<string> => {
    try {
      const request = new DeleteVersionRequest();
      request.setVersionId(versionId);

      const response = await mapcodeServiceClient.deleteVersion(request, getAuthMetadata());
      const status = response.getStatus();

      return status?.getMessage() || 'Version deleted successfully';
    } catch (error) {
      const errorMessage = handleGrpcError(error as RpcError);
      throw new Error(errorMessage);
    }
  },

  /**
   * Translate a single question code
   * Dịch một mã câu hỏi
   */
  translateCode: async (questionCode: string): Promise<MapCodeTranslationData> => {
    try {
      const request = new TranslateCodeRequest();
      request.setQuestionCode(questionCode);

      const response = await mapcodeServiceClient.translateCode(request, getAuthMetadata());
      const translation = response.getTranslation();

      if (!translation) {
        throw new Error('No translation returned from server');
      }

      return mapTranslationFromPb(translation);
    } catch (error) {
      const errorMessage = handleGrpcError(error as RpcError);
      throw new Error(errorMessage);
    }
  },

  /**
   * Translate multiple question codes
   * Dịch nhiều mã câu hỏi
   */
  translateCodes: async (questionCodes: string[]): Promise<Record<string, string>> => {
    try {
      const request = new TranslateCodesRequest();
      request.setQuestionCodesList(questionCodes);

      const response = await mapcodeServiceClient.translateCodes(request, getAuthMetadata());
      const translationsMap = response.getTranslationsMap();

      const result: Record<string, string> = {};
      translationsMap.forEach((value: string, key: string) => {
        result[key] = value;
      });

      return result;
    } catch (error) {
      const errorMessage = handleGrpcError(error as RpcError);
      throw new Error(errorMessage);
    }
  },

  /**
   * Get hierarchy navigation for a question code
   * Lấy cấu trúc phân cấp cho mã câu hỏi
   */
  getHierarchyNavigation: async (questionCode: string): Promise<HierarchyNavigationData> => {
    try {
      const request = new GetHierarchyNavigationRequest();
      request.setQuestionCode(questionCode);

      const response = await mapcodeServiceClient.getHierarchyNavigation(request, getAuthMetadata());
      const navigation = response.getNavigation();

      if (!navigation) {
        throw new Error('No navigation returned from server');
      }

      return mapHierarchyNavigationFromPb(navigation);
    } catch (error) {
      const errorMessage = handleGrpcError(error as RpcError);
      throw new Error(errorMessage);
    }
  },

  /**
   * Get storage information
   * Lấy thông tin lưu trữ
   */
  getStorageInfo: async (): Promise<StorageInfoData> => {
    try {
      const request = new GetStorageInfoRequest();

      const response = await mapcodeServiceClient.getStorageInfo(request, getAuthMetadata());
      const storage = response.getStorage();

      if (!storage) {
        throw new Error('No storage info returned from server');
      }

      return mapStorageInfoFromPb(storage);
    } catch (error) {
      const errorMessage = handleGrpcError(error as RpcError);
      throw new Error(errorMessage);
    }
  },
};

export default MapCodeService;

File: apps/backend/internal/grpc/mapcode_service.go
package grpc

import (
	"context"
	"fmt"

	mapcode "exam-bank-system/apps/backend/internal/service/content/mapcode"
	"exam-bank-system/apps/backend/pkg/proto/common"
	pb "exam-bank-system/apps/backend/pkg/proto/v1"
	"google.golang.org/protobuf/types/known/timestamppb"
)

// MapCodeServiceServer implements the MapCode gRPC service
type MapCodeServiceServer struct {
	pb.UnimplementedMapCodeServiceServer
	mapCodeMgmt *mapcode.MapCodeMgmt
}

// NewMapCodeServiceServer creates a new MapCode service server
func NewMapCodeServiceServer(mapCodeMgmt *mapcode.MapCodeMgmt) *MapCodeServiceServer {
	return &MapCodeServiceServer{
		mapCodeMgmt: mapCodeMgmt,
	}
}

// CreateVersion creates a new MapCode version
func (s *MapCodeServiceServer) CreateVersion(ctx context.Context, req *pb.CreateVersionRequest) (*pb.CreateVersionResponse, error) {
	// Validate request
	if req.Version == "" {
		return &pb.CreateVersionResponse{
			Status: &common.Response{
				Success: false,
				Message: "Version identifier is required",
			},
		}, nil
	}

	if req.Name == "" {
		return &pb.CreateVersionResponse{
			Status: &common.Response{
				Success: false,
				Message: "Version name is required",
			},
		}, nil
	}

	// Create version
	version, err := s.mapCodeMgmt.CreateVersion(ctx, req.Version, req.Name, req.Description, req.CreatedBy)
	if err != nil {
		return &pb.CreateVersionResponse{
			Status: &common.Response{
				Success: false,
				Message: fmt.Sprintf("Failed to create version: %v", err),
			},
		}, nil
	}

	// Convert to proto
	protoVersion := &pb.MapCodeVersion{
		Id:          version.ID.String,
		Version:     version.Version.String,
		Name:        version.Name.String,
		Description: version.Description.String,
		FilePath:    version.FilePath.String,
		IsActive:    version.IsActive.Bool,
		CreatedBy:   version.CreatedBy.String,
		CreatedAt:   timestamppb.New(version.CreatedAt.Time),
		UpdatedAt:   timestamppb.New(version.UpdatedAt.Time),
	}

	return &pb.CreateVersionResponse{
		Status: &common.Response{
			Success: true,
			Message: "Version created successfully",
		},
		Version: protoVersion,
	}, nil
}

// GetVersions retrieves all versions with pagination
func (s *MapCodeServiceServer) GetVersions(ctx context.Context, req *pb.GetVersionsRequest) (*pb.GetVersionsResponse, error) {
	// Set default pagination
	page := req.Page
	if page <= 0 {
		page = 1
	}
	pageSize := req.PageSize
	if pageSize <= 0 {
		pageSize = 10
	}

	offset := (page - 1) * pageSize

	// Get versions
	versions, err := s.mapCodeMgmt.GetAllVersions(ctx, int(offset), int(pageSize))
	if err != nil {
		return &pb.GetVersionsResponse{
			Status: &common.Response{
				Success: false,
				Message: fmt.Sprintf("Failed to get versions: %v", err),
			},
		}, nil
	}

	// Convert to proto
	var protoVersions []*pb.MapCodeVersion
	for _, version := range versions {
		protoVersions = append(protoVersions, &pb.MapCodeVersion{
			Id:          version.ID.String,
			Version:     version.Version.String,
			Name:        version.Name.String,
			Description: version.Description.String,
			FilePath:    version.FilePath.String,
			IsActive:    version.IsActive.Bool,
			CreatedBy:   version.CreatedBy.String,
			CreatedAt:   timestamppb.New(version.CreatedAt.Time),
			UpdatedAt:   timestamppb.New(version.UpdatedAt.Time),
		})
	}

	// TODO: Implement proper pagination count
	pagination := &common.PaginationResponse{
		Page:       page,
		Limit:      pageSize,
		TotalPages: 1, // Calculate based on total count
		TotalCount: int32(len(protoVersions)),
	}

	return &pb.GetVersionsResponse{
		Status: &common.Response{
			Success: true,
			Message: "Versions retrieved successfully",
		},
		Versions:   protoVersions,
		Pagination: pagination,
	}, nil
}

// GetActiveVersion retrieves the currently active version
func (s *MapCodeServiceServer) GetActiveVersion(ctx context.Context, req *pb.GetActiveVersionRequest) (*pb.GetActiveVersionResponse, error) {
	version, err := s.mapCodeMgmt.GetActiveVersion(ctx)
	if err != nil {
		return &pb.GetActiveVersionResponse{
			Status: &common.Response{
				Success: false,
				Message: fmt.Sprintf("No active version found: %v", err),
			},
		}, nil
	}

	protoVersion := &pb.MapCodeVersion{
		Id:          version.ID.String,
		Version:     version.Version.String,
		Name:        version.Name.String,
		Description: version.Description.String,
		FilePath:    version.FilePath.String,
		IsActive:    version.IsActive.Bool,
		CreatedBy:   version.CreatedBy.String,
		CreatedAt:   timestamppb.New(version.CreatedAt.Time),
		UpdatedAt:   timestamppb.New(version.UpdatedAt.Time),
	}

	return &pb.GetActiveVersionResponse{
		Status: &common.Response{
			Success: true,
			Message: "Active version retrieved successfully",
		},
		Version: protoVersion,
	}, nil
}

// SetActiveVersion sets a version as active
func (s *MapCodeServiceServer) SetActiveVersion(ctx context.Context, req *pb.SetActiveVersionRequest) (*pb.SetActiveVersionResponse, error) {
	if req.VersionId == "" {
		return &pb.SetActiveVersionResponse{
			Status: &common.Response{
				Success: false,
				Message: "Version ID is required",
			},
		}, nil
	}

	err := s.mapCodeMgmt.SetActiveVersion(ctx, req.VersionId)
	if err != nil {
		return &pb.SetActiveVersionResponse{
			Status: &common.Response{
				Success: false,
				Message: fmt.Sprintf("Failed to set active version: %v", err),
			},
		}, nil
	}

	// Clear cache after version switch
	s.mapCodeMgmt.ClearCache()

	return &pb.SetActiveVersionResponse{
		Status: &common.Response{
			Success: true,
			Message: "Active version updated successfully",
		},
		Message: "Version activated and cache cleared",
	}, nil
}

// DeleteVersion deletes a version (only if not active)
func (s *MapCodeServiceServer) DeleteVersion(ctx context.Context, req *pb.DeleteVersionRequest) (*pb.DeleteVersionResponse, error) {
	if req.VersionId == "" {
		return &pb.DeleteVersionResponse{
			Status: &common.Response{
				Success: false,
				Message: "Version ID is required",
			},
		}, nil
	}

	err := s.mapCodeMgmt.DeleteVersion(ctx, req.VersionId)
	if err != nil {
		return &pb.DeleteVersionResponse{
			Status: &common.Response{
				Success: false,
				Message: fmt.Sprintf("Failed to delete version: %v", err),
			},
		}, nil
	}

	return &pb.DeleteVersionResponse{
		Status: &common.Response{
			Success: true,
			Message: "Version deleted successfully",
		},
		Message: "Version and associated files removed",
	}, nil
}

// TranslateCode translates a single question code
func (s *MapCodeServiceServer) TranslateCode(ctx context.Context, req *pb.TranslateCodeRequest) (*pb.TranslateCodeResponse, error) {
	if req.QuestionCode == "" {
		return &pb.TranslateCodeResponse{
			Status: &common.Response{
				Success: false,
				Message: "Question code is required",
			},
		}, nil
	}

	translation, err := s.mapCodeMgmt.TranslateQuestionCode(ctx, req.QuestionCode)
	if err != nil {
		return &pb.TranslateCodeResponse{
			Status: &common.Response{
				Success: false,
				Message: fmt.Sprintf("Translation failed: %v", err),
			},
		}, nil
	}

	return &pb.TranslateCodeResponse{
		Status: &common.Response{
			Success: true,
			Message: "Translation completed successfully",
		},
		Translation: &pb.MapCodeTranslation{
			QuestionCode: req.QuestionCode,
			Translation:  translation,
			// TODO: Add individual parsed components
		},
	}, nil
}

// TranslateCodes translates multiple question codes
func (s *MapCodeServiceServer) TranslateCodes(ctx context.Context, req *pb.TranslateCodesRequest) (*pb.TranslateCodesResponse, error) {
	if len(req.QuestionCodes) == 0 {
		return &pb.TranslateCodesResponse{
			Status: &common.Response{
				Success: false,
				Message: "At least one question code is required",
			},
		}, nil
	}

	translations, err := s.mapCodeMgmt.TranslateQuestionCodes(ctx, req.QuestionCodes)
	if err != nil {
		return &pb.TranslateCodesResponse{
			Status: &common.Response{
				Success: false,
				Message: fmt.Sprintf("Batch translation failed: %v", err),
			},
		}, nil
	}

	return &pb.TranslateCodesResponse{
		Status: &common.Response{
			Success: true,
			Message: fmt.Sprintf("Translated %d question codes", len(translations)),
		},
		Translations: translations,
	}, nil
}

// GetHierarchyNavigation returns navigation structure for a question code
func (s *MapCodeServiceServer) GetHierarchyNavigation(ctx context.Context, req *pb.GetHierarchyNavigationRequest) (*pb.GetHierarchyNavigationResponse, error) {
	if req.QuestionCode == "" {
		return &pb.GetHierarchyNavigationResponse{
			Status: &common.Response{
				Success: false,
				Message: "Question code is required",
			},
		}, nil
	}

	navigation, err := s.mapCodeMgmt.GetHierarchyNavigation(ctx, req.QuestionCode)
	if err != nil {
		return &pb.GetHierarchyNavigationResponse{
			Status: &common.Response{
				Success: false,
				Message: fmt.Sprintf("Failed to get hierarchy navigation: %v", err),
			},
		}, nil
	}

	// Convert to proto
	protoNavigation := &pb.HierarchyNavigation{
		QuestionCode: navigation.QuestionCode,
		Breadcrumbs:  navigation.Breadcrumbs,
	}

	if navigation.Grade != nil {
		protoNavigation.Grade = &pb.HierarchyLevel{
			Code:        navigation.Grade.Code,
			Name:        navigation.Grade.Name,
			Description: navigation.Grade.Description,
		}
	}

	if navigation.Subject != nil {
		protoNavigation.Subject = &pb.HierarchyLevel{
			Code:        navigation.Subject.Code,
			Name:        navigation.Subject.Name,
			Description: navigation.Subject.Description,
		}
	}

	if navigation.Chapter != nil {
		protoNavigation.Chapter = &pb.HierarchyLevel{
			Code:        navigation.Chapter.Code,
			Name:        navigation.Chapter.Name,
			Description: navigation.Chapter.Description,
		}
	}

	if navigation.Level != nil {
		protoNavigation.Level = &pb.HierarchyLevel{
			Code:        navigation.Level.Code,
			Name:        navigation.Level.Name,
			Description: navigation.Level.Description,
		}
	}

	if navigation.Lesson != nil {
		protoNavigation.Lesson = &pb.HierarchyLevel{
			Code:        navigation.Lesson.Code,
			Name:        navigation.Lesson.Name,
			Description: navigation.Lesson.Description,
		}
	}

	if navigation.Form != nil {
		protoNavigation.Form = &pb.HierarchyLevel{
			Code:        navigation.Form.Code,
			Name:        navigation.Form.Name,
			Description: navigation.Form.Description,
		}
	}

	return &pb.GetHierarchyNavigationResponse{
		Status: &common.Response{
			Success: true,
			Message: "Hierarchy navigation retrieved successfully",
		},
		Navigation: protoNavigation,
	}, nil
}

// GetStorageInfo returns storage information
func (s *MapCodeServiceServer) GetStorageInfo(ctx context.Context, req *pb.GetStorageInfoRequest) (*pb.GetStorageInfoResponse, error) {
	storageInfo, err := s.mapCodeMgmt.GetStorageInfo(ctx)
	if err != nil {
		return &pb.GetStorageInfoResponse{
			Status: &common.Response{
				Success: false,
				Message: fmt.Sprintf("Failed to get storage info: %v", err),
			},
		}, nil
	}

	protoStorage := &pb.StorageInfo{
		TotalVersions:  int32(storageInfo.TotalVersions),
		MaxVersions:    int32(storageInfo.MaxVersions),
		AvailableSlots: int32(storageInfo.AvailableSlots),
		WarningLevel:   int32(storageInfo.WarningLevel),
		IsNearLimit:    storageInfo.IsNearLimit,
		IsAtLimit:      storageInfo.IsAtLimit,
		WarningMessage: storageInfo.GetWarningMessage(),
	}

	return &pb.GetStorageInfoResponse{
		Status: &common.Response{
			Success: true,
			Message: "Storage information retrieved successfully",
		},
		Storage: protoStorage,
	}, nil
}

// GetMapCodeConfig returns the full MapCode configuration
func (s *MapCodeServiceServer) GetMapCodeConfig(ctx context.Context, req *pb.GetMapCodeConfigRequest) (*pb.GetMapCodeConfigResponse, error) {
	config, err := s.mapCodeMgmt.GetMapCodeConfig(ctx, req.VersionId)
	if err != nil {
		return &pb.GetMapCodeConfigResponse{
			Status: &common.Response{
				Success: false,
				Message: fmt.Sprintf("Failed to get MapCode config: %v", err),
			},
		}, nil
	}

	// Convert config maps to proto format
	protoConfig := &pb.MapCodeConfig{
		Version:  config.Version,
		Grades:   config.Grades,
		Subjects: config.Subjects,
		Chapters: config.Chapters,
		Levels:   config.Levels,
		Lessons:  config.Lessons,
		Forms:    config.Forms,
	}

	return &pb.GetMapCodeConfigResponse{
		Status: &common.Response{
			Success: true,
			Message: "MapCode configuration retrieved successfully",
		},
		Config: protoConfig,
	}, nil
}

